src/org/sosy_lab/cpachecker/cfa/mutation/strategy/GenericCFAMutationStrategy.java

```diff
@@ -0,0 +1,368 @@
+/*
+ *  CPAchecker is a tool for configurable software verification.
+ *  This file is part of CPAchecker.
+ *
+ *  Copyright (C) 2007-2020  Dirk Beyer
+ *  All rights reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.sosy_lab.cpachecker.cfa.mutation.strategy;
+
+import java.io.PrintStream;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.sosy_lab.common.log.LogManager;
+import org.sosy_lab.cpachecker.cfa.ParseResult;
+import org.sosy_lab.cpachecker.core.interfaces.Statistics;
+import org.sosy_lab.cpachecker.util.statistics.StatCounter;
+import org.sosy_lab.cpachecker.util.statistics.StatInt;
+import org.sosy_lab.cpachecker.util.statistics.StatKind;
+import org.sosy_lab.cpachecker.util.statistics.StatisticsWriter;
+
+public abstract class GenericCFAMutationStrategy<ObjectKey, RollbackInfo>
+    extends AbstractCFAMutationStrategy {
+
+  private Collection<ObjectKey> objectsBefore;
+
+  private Deque<ObjectKey> currentMutation = new ArrayDeque<>();
+  private Deque<RollbackInfo> rollbackInfos = new ArrayDeque<>();
+  private final Set<ObjectKey> previousPasses = new HashSet<>();
+  private final Set<ObjectKey> previousMutations = new HashSet<>();
+  private int rate = 0;
+  private int depth = -1;
+  private int levelSize = -1;
+  private int batchNum = -1;
+  private int batchCount = -1;
+  private int batchSize = -1;
+  private final boolean tryAllAtFirst;
```

Comment:
In some cases it is possible remove everything in the first go, so we set tryAllAtFirst=true; otherwise it is set to false. It depends on pRate in constructor. Both pRate and tryAllAtFirst can be removed for simplification.